File: data_baseline/openrouter__qwen__qwen3-coder-plus/javascript/rep3/test_server.js
const { spawn } = require('child_process');
const path = require('path');

// Test script to verify our API implementation

async function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function testImplementation() {
    console.log('Starting server tests...');

    // Start the server on a random available port
    const testPort = 4567;
    const env = { ...process.env, PORT: testPort.toString() };
    const serverProcess = spawn('node', ['server.js', '--port', testPort.toString()], { cwd: process.cwd() });

    let serverStarted = false;
    serverProcess.stdout.on('data', (data) => {
        if (data.toString().includes(`Server running on 0.0.0.0:${testPort}`)) {
            serverStarted = true;
            console.log('✓ Server started successfully on port:', testPort);
        }
        console.log(`Server output: ${data}`);
    });

    serverProcess.stderr.on('data', (data) => {
        console.error(`Server error: ${data}`);
    });

    // Wait for server to start
    while (!serverStarted) {
        await sleep(100);
    }
    await sleep(100); // Additional wait for port to be fully bound

    // Import here after server starts since we need the modules
    const http = require('http');
    const { once } = require('events');

    async function makeRequest(options, postData = null) {
        return new Promise((resolve, reject) => {
            const req = http.request(options, (res) => {
                let data = '';
                res.on('data', (chunk) => {
                    data += chunk;
                });
                res.on('end', () => {
                    try {
                        const result = { 
                            statusCode: res.statusCode, 
                            headers: res.headers,
                            body: data ? JSON.parse(data) : null
                        };
                        resolve(result);
                    } catch (e) {
                        resolve({ 
                            statusCode: res.statusCode, 
                            headers: res.headers,
                            body: data  // Return raw data if not JSON
                        });
                    }
                });
            });

            req.on('error', (e) => {
                reject(e);
            });

            if (postData) {
                req.write(postData);
            }
            
            req.end();
        });
    }

    // Test variables to hold state between tests
    let user1SessionCookie = null;
    let user2SessionCookie = null;
    let user1Credentials = { username: 'testuser1', password: 'password123' };
    let user2Credentials = { username: 'testuser2', password: 'password456' };
    let createdTodoId = null;

    try {
        console.log('\n🧪 Testing Registration...');
        
        // Test registration of user 1
        let response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/register',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }, JSON.stringify(user1Credentials));
        
        if (response.statusCode === 201 && response.body.id && response.body.username === user1Credentials.username) {
            console.log('✓ Registration success for user 1');
        } else {
            console.log('✗ Registration failed for user 1:', response);
            throw new Error('Registration Test Failed');
        }

        // Test registration of user 2
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/register',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }, JSON.stringify(user2Credentials));
        
        if (response.statusCode === 201 && response.body.id && response.body.username === user2Credentials.username) {
            console.log('✓ Registration success for user 2');
        } else {
            console.log('✗ Registration failed for user 2:', response);
            throw new Error('Registration Test Failed');
        }

        // Test registration conflict
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/register',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }, JSON.stringify(user1Credentials));  // Try to re-register same user
        
        if (response.statusCode === 409 && response.body.error === 'Username already exists') {
            console.log('✓ Conflict handling success for duplicate username');
        } else {
            console.log('✗ Conflict test failed:', response);
            throw new Error('Conflict Test Failed');
        }

        console.log('\n🧪 Testing Login...');
        
        // Test login for user 1
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/login',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }, JSON.stringify(user1Credentials));
        
        if (response.statusCode === 200 && response.headers['set-cookie'] && 
            response.body.id && response.body.username === user1Credentials.username) {
            user1SessionCookie = response.headers['set-cookie'][0].split(';')[0];  // Extract cookie
            console.log('✓ Login success for user 1 - Session:', user1SessionCookie);
        } else {
            console.log('✗ Login failed for user 1:', response);
            throw new Error('Login Test Failed');
        }

        // Test login for user 2
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/login',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }, JSON.stringify(user2Credentials));
        
        if (response.statusCode === 200 && response.headers['set-cookie'] &&
            response.body.id && response.body.username === user2Credentials.username) {
            user2SessionCookie = response.headers['set-cookie'][0].split(';')[0];  // Extract cookie
            console.log('✓ Login success for user 2 - Session:', user2SessionCookie);
        } else {
            console.log('✗ Login failed for user 2:', response);
            throw new Error('Login Test Failed');
        }

        // Test invalid credentials
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/login',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            }
        }, JSON.stringify({ username: user1Credentials.username, password: 'wrongpass' }));
        
        if (response.statusCode === 401 && response.body.error === 'Invalid credentials') {
            console.log('✓ Invalid credentials handled correctly');
        } else {
            console.log('✗ Invalid credentials test failed:', response);
            throw new Error('Invalid Credentials Test Failed');
        }

        console.log('\n🧪 Testing Authentication Required Routes...');
        
        // Test /me without authentication
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/me',
            method: 'GET'
        });
        
        if (response.statusCode === 401 && response.body.error === 'Authentication required') {
            console.log('✓ Auth required enforced on /me');
        } else {
            console.log('✗ Auth required test failed on /me:', response);
            throw new Error('Auth Required Test Failed');
        }

        // Test protected routes with auth (get user info)
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/me',
            method: 'GET',
            headers: {
                'Cookie': user1SessionCookie
            }
        });
        
        if (response.statusCode === 200 && response.body.username === user1Credentials.username) {
            console.log('✓ /me returns correct user info');
        } else {
            console.log('✗ /me test failed:', response);
            throw new Error('/me Test Failed');
        }

        console.log('\n🧪 Testing Todo Creation & Retrieval...');
        
        // Test creating a todo
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/todos',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': user1SessionCookie
            }
        }, JSON.stringify({
            title: 'Test Todo',
            description: 'This is a test todo item'
        }));
        
        if (response.statusCode === 201 && response.body.title === 'Test Todo' && 
            response.body.description === 'This is a test todo item' && 
            response.body.completed === false) {
            createdTodoId = response.body.id;
            console.log('✓ Todo creation successful - ID:', createdTodoId);
        } else {
            console.log('✗ Todo creation failed:', response);
            throw new Error('Todo Creation Test Failed');
        }

        // Test creating todo with empty title (should fail)
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/todos',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': user1SessionCookie
            }
        }, JSON.stringify({
            title: '',
            description: 'This should fail'
        }));
        
        if (response.statusCode === 400 && response.body.error === 'Title is required') {
            console.log('✓ Todo creation validation works');
        } else {
            console.log('✗ Todo validation test failed:', response);
            throw new Error('Todo Validation Test Failed');
        }

        // Test getting user todos
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/todos',
            method: 'GET',
            headers: {
                'Cookie': user1SessionCookie
            }
        });
        
        if (response.statusCode === 200 && Array.isArray(response.body) && response.body.length > 0) {
            console.log('✓ Todo retrieval successful - Count:', response.body.length);
        } else {
            console.log('✗ Todo retrieval failed:', response);
            throw new Error('Todo Retrieval Test Failed');
        }

        console.log('\n🧪 Testing Todo CRUD Operations...');
        
        // Test getting a specific todo
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: `/todos/${createdTodoId}`,
            method: 'GET',
            headers: {
                'Cookie': user1SessionCookie
            }
        });
        
        if (response.statusCode === 200 && response.body.id === createdTodoId) {
            console.log('✓ Get specific todo successful');
        } else {
            console.log('✗ Get specific todo failed:', response);
            throw new Error('Get Todo Test Failed');
        }

        // Test updating a specific todo
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: `/todos/${createdTodoId}`,
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': user1SessionCookie
            }
        }, JSON.stringify({
            title: 'Updated Title',
            completed: true
        }));
        
        if (response.statusCode === 200 && response.body.title === 'Updated Title' && 
            response.body.completed === true && response.body.id === createdTodoId) {
            console.log('✓ Todo update successful');
        } else {
            console.log('✗ Todo update failed:', response);
            throw new Error('Todo Update Test Failed');
        }

        // Test updating with empty title (should fail)
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: `/todos/${createdTodoId}`,
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': user1SessionCookie
            }
        }, JSON.stringify({
            title: ''
        }));
        
        if (response.statusCode === 400 && response.body.error === 'Title is required') {
            console.log('✓ Todo update validation works');
        } else {
            console.log('✗ Todo update validation test failed:', response);
            throw new Error('Todo Update Validation Test Failed');
        }

        // Test accessing another user's todo (should fail with 404)
        const user2Response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/todos',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': user2SessionCookie
            }
        }, JSON.stringify({
            title: 'User2 Todo'
        }));
        
        if (user2Response.statusCode === 201 && user2Response.body.title === 'User2 Todo') {
            const user2TodoId = user2Response.body.id;
            console.log('✓ User2 todo created - ID:', user2TodoId);
            
            // Now test that user1 can't access user2's todo
            response = await makeRequest({
                hostname: 'localhost',
                port: testPort,
                path: `/todos/${user2TodoId}`,
                method: 'GET',
                headers: {
                    'Cookie': user1SessionCookie  // User1 trying to access User2's todo
                }
            });
            
            if (response.statusCode === 404 && response.body.error === 'Todo not found') {
                console.log('✓ Cross-user protection successful (get)');
            } else {
                console.log('✗ Cross-user protection failed (get):', response);
                throw new Error('Cross-user Protection Test Failed');
            }
            
            // Also test update
            response = await makeRequest({
                hostname: 'localhost',
                port: testPort,
                path: `/todos/${user2TodoId}`,
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    'Cookie': user1SessionCookie  // User1 trying to update User2's todo
                },
                body: JSON.stringify({ title: 'Hacked title' })
            });
            
            if (response.statusCode === 404 && response.body.error === 'Todo not found') {
                console.log('✓ Cross-user protection successful (update)');
            } else {
                console.log('✗ Cross-user protection failed (update):', response);
                throw new Error('Cross-user Protection Test Failed');
            }
            
        } else {
            console.log('✗ User2 todo creation failed:', user2Response);
            throw new Error('User2 Todo Creation Failed');
        }

        // Test deleting a todo
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: `/todos/${createdTodoId}`,
            method: 'DELETE',
            headers: {
                'Cookie': user1SessionCookie
            }
        });
        
        if (response.statusCode === 204) {
            console.log('✓ Todo deletion successful');
        } else {
            console.log('✗ Todo deletion failed:', response);
            throw new Error('Todo Deletion Test Failed');
        }

        // Test that deleted todo doesn't exist anymore
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: `/todos/${createdTodoId}`,
            method: 'GET',
            headers: {
                'Cookie': user1SessionCookie
            }
        });
        
        if (response.statusCode === 404 && response.body.error === 'Todo not found') {
            console.log('✓ Deleted todo properly inaccessible');
        } else {
            console.log('✗ Deleted todo still accessible:', response);
            throw new Error('Deleted Todo Accessibility Test Failed');
        }

        console.log('\n🧪 Testing Password Change...');
        
        // Test changing password
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/password',
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Cookie': user1SessionCookie
            }
        }, JSON.stringify({
            old_password: user1Credentials.password,
            new_password: 'newpassword123'
        }));
        
        if (response.statusCode === 200) {
            console.log('✓ Password change successful');
            
            // Now try to login with new password
            response = await makeRequest({
                hostname: 'localhost',
                port: testPort,
                path: '/login',
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                }
            }, JSON.stringify({...user1Credentials, password: 'newpassword123'}));
            
            if (response.statusCode === 200) {
                console.log('✓ New password login successful');
                
                // Update session for later tests
                user1SessionCookie = response.headers['set-cookie'][0].split(';')[0];
            } else {
                console.log('✗ New password login failed:', response);
                throw new Error('New Password Login Test Failed');
            }
        } else {
            console.log('✗ Password change failed:', response);
            throw new Error('Password Change Test Failed');
        }

        console.log('\n🧪 Testing Logout...');
        
        // Test logout
        response = await makeRequest({
            hostname: 'localhost',
            port: testPort,
            path: '/logout',
            method: 'POST',
            headers: {
                'Cookie': user1SessionCookie
            }
        });
        
        if (response.statusCode === 200) {
            console.log('✓ Logout successful');
            
            // Verify session invalidation by attempting to access protected resource
            response = await makeRequest({
                hostname: 'localhost',
                port: testPort,
                path: '/me',
                method: 'GET',
                headers: {
                    'Cookie': user1SessionCookie  // Using invalidated session
                }
            });
            
            if (response.statusCode === 401 && response.body.error === 'Authentication required') {
                console.log('✓ Session properly invalidated after logout');
            } else {
                console.log('✗ Session not invalidated after logout:', response);
                throw new Error('Session Invalidating After Logout Test Failed');
            }
        } else {
            console.log('✗ Logout failed:', response);
            throw new Error('Logout Test Failed');
        }

        console.log('\n🎉 All tests passed!');
        
    } catch (error) {
        console.error('\n💥 Test failed:', error.message);
        serverProcess.kill();
        process.exit(1);
    }

    // Clean up: Kill the server process
    serverProcess.kill();
    console.log('\n✅ Server terminated.');
}

testImplementation();

File: data_baseline/openrouter__qwen__qwen3-coder-plus/python/rep3/test_server.py
#!/usr/bin/env python3
"""
Test script for the Todo app server.
This script tests all endpoints to ensure they're functioning properly.
"""

import subprocess
import time
import requests
import json
import os
import signal


def wait_for_server(url, timeout=30):
    """Wait for server to be ready."""
    start_time = time.time() 
    while time.time() - start_time < timeout:
        try:
            response = requests.get(f"http://localhost:8000/me", allow_redirects=False)
            # Server might return 401 (not 404) if it's running
            if response.status_code in [200, 401]:
                return True
        except requests.exceptions.ConnectionError:
            pass
        time.sleep(0.5)
    return False


def reset_cookies(session, initial_cookies=None):
    """Helper to keep existing session cookies but add new ones."""
    if initial_cookies is None:
        initial_cookies = {}
    
    # Copy over existing cookies from session
    for cookie in session.cookies:
        initial_cookies[cookie.name] = cookie.value
    session.cookies.clear()
    session.cookies.update(initial_cookies)
    return session


def main():
    print("Starting Todo App server tests...")
    
    # Start the server
    process = subprocess.Popen(['python3', 'server.py', '--port', '8000'])
    
    # Wait for server to be reachable
    print("Waiting for server to start...")
    if not wait_for_server("http://localhost:8000"):
        print("ERROR: Server failed to start")
        process.terminate()
        return
    
    print("Server is running!")
    
    # Create a session to maintain cookies
    session = requests.Session()
    
    print("\n=== Testing REGISTER Endpoint ===")
    # Test register new user
    res = session.post("http://localhost:8000/register", 
                      json={"username": "testuser", "password": "password123"})
    assert res.status_code == 201
    assert res.json()["username"] == "testuser"
    print("✓ Register successful")
    
    # Test register with invalid username (too short)
    res = session.post("http://localhost:8000/register", 
                      json={"username": "ab", "password": "password123"})
    assert res.status_code == 400
    assert "Invalid username" in res.json()["error"]
    print("✓ Invalid username validation works")
    
    # Test register with invalid username (invalid characters)
    res = session.post("http://localhost:8000/register", 
                      json={"username": "test@user", "password": "password123"})
    assert res.status_code == 400
    assert "Invalid username" in res.json()["error"]
    print("✓ Invalid character username validation works")
    
    # Test register with duplicate username
    res = session.post("http://localhost:8000/register", 
                      json={"username": "testuser", "password": "password123"})
    assert res.status_code == 409
    assert "Username already exists" in res.json()["error"]
    print("✓ Duplicate username validation works")
    
    # Test register with short password
    res = session.post("http://localhost:8000/register", 
                      json={"username": "testuser2", "password": "pass"})
    assert res.status_code == 400
    assert "Password too short" in res.json()["error"]
    print("✓ Short password validation works")
    
    print("\n=== Testing LOGIN Endpoint ===")
    # Clear session to test new login properly
    session.cookies.clear()
    # Test successful login
    res = session.post("http://localhost:8000/login", 
                      json={"username": "testuser", "password": "password123"})
    assert res.status_code == 200
    assert res.json()["username"] == "testuser"
    assert "session_id" in [c.name for c in session.cookies]
    print("✓ Login successful")
    
    # Test incorrect login
    session.cookies.clear()  # Clear session to start fresh
    s = requests.Session()
    res = s.post("http://localhost:8000/login", 
                 json={"username": "testuser", "password": "wrongpassword"})
    assert res.status_code == 401
    assert "Invalid credentials" in res.json()["error"]
    print("✓ Incorrect login validation works")
    
    # Test non-existent user login
    res = s.post("http://localhost:8000/login", 
                 json={"username": "nonexistent", "password": "password123"})
    assert res.status_code == 401
    assert "Invalid credentials" in res.json()["error"]
    print("✓ Non-existent user login validation works")
    
    print("\n=== Testing PROTECTED ENDPOINTS without Auth ===")
    # Now test authenticated endpoints WITHOUT proper auth
    noauth_session = requests.Session()
    
    # Get me without auth
    res = noauth_session.get("http://localhost:8000/me")
    assert res.status_code == 401
    assert "Authentication required" in res.json()["error"]
    print("✓ Get '/me' without auth correctly returns 401")
    
    # Create todo without auth
    res = noauth_session.post("http://localhost:8000/todos", 
                             json={"title": "Test todo", "description": "A test todo"})
    assert res.status_code == 401
    assert "Authentication required" in res.json()["error"]
    print("✓ Create '/todos' without auth correctly returns 401")
    
    print("\n=== Testing ME Endpoint ===")
    # Test Get Me
    res = session.get("http://localhost:8000/me")
    assert res.status_code == 200
    assert res.json()["username"] == "testuser"
    print("✓ Get '/me' successful")
    
    print("\n=== Testing TODO Endpoints ===")
    # Create first todo
    res = session.post("http://localhost:8000/todos", 
                       json={"title": "First todo", "description": "My first task"})
    assert res.status_code == 201
    assert res.json()["title"] == "First todo"
    todo1_id = res.json()["id"]
    print("✓ Created first todo")
    
    # Create another todo
    res = session.post("http://localhost:8000/todos", 
                       json={"title": "Second todo", "description": ""})
    assert res.status_code == 201
    assert res.json()["title"] == "Second todo"
    assert res.json()["description"] == ""
    todo2_id = res.json()["id"]
    print("✓ Created second todo")
    
    # Create todo without title (should fail)
    res = session.post("http://localhost:8000/todos", 
                       json={"description": "todo without title"})
    assert res.status_code == 400
    assert "Title is required" in res.json()["error"]
    print("✓ Validation of missing title works")
    
    # Get all todos
    res = session.get("http://localhost:8000/todos")
    assert res.status_code == 200
    todos = res.json()
    assert len(todos) == 2
    assert todos[0]["id"] <= todos[1]["id"]  # Ensure correct ordering by ID
    print("✓ Get all todos successful")
    
    # Get specific todo
    res = session.get(f"http://localhost:8000/todos/{todo1_id}")
    assert res.status_code == 200
    assert res.json()["id"] == todo1_id
    assert res.json()["title"] == "First todo"
    print("✓ Get specific todo successful")
    
    # Test get non-existent todo
    res = session.get("http://localhost:8000/todos/9999")
    assert res.status_code == 404
    assert "Todo not found" in res.json()["error"]
    print("✓ Get non-existent todo returns 404")
    
    # Update todo partially
    res = session.put(f"http://localhost:8000/todos/{todo1_id}", 
                     json={"completed": True})
    assert res.status_code == 200
    assert res.json()["completed"] == True
    print("✓ Partially update todo successful")
    
    # Verify update worked
    res = session.get(f"http://localhost:8000/todos/{todo1_id}")
    assert res.json()["completed"] == True
    print("✓ Updated state confirmed")
    
    # Test invalid update (empty title)
    res = session.put(f"http://localhost:8000/todos/{todo1_id}", 
                     json={"title": ""})
    assert res.status_code == 400
    assert "Title is required" in res.json()["error"]
    print("✓ Update with empty title validation works")
    
    # Test updating another property
    res = session.put(f"http://localhost:8000/todos/{todo1_id}", 
                     json={"title": "Updated first todo", "description": "Updated description"})
    assert res.status_code == 200
    assert res.json()["title"] == "Updated first todo"
    assert res.json()["description"] == "Updated description"
    print("✓ Full update todo successful")
    
    # Try to access another user's todo (after creating one for that user)
    # For our simple case, we'll test that the same user can't access non-existent todo of someone else's ID space
    # but instead let's confirm user isolation via a different approach
    # First logout
    res = session.post("http://localhost:8000/logout")
    assert res.status_code == 200
    print("✓ Logout successful")
    
    # Clear cookies between sessions
    session.cookies.clear()
    
    # Now register another user
    res = session.post("http://localhost:8000/register", 
                      json={"username": "otheruser", "password": "password123"})
    assert res.status_code == 201
    print("✓ Other user registration successful")
    
    # Login with other user
    res = session.post("http://localhost:8000/login", 
                      json={"username": "otheruser", "password": "password123"})
    assert res.status_code == 200
    print("✓ Other user login successful")
    
    # Try to access the first user's todo using the other user's session
    res = session.get(f"http://localhost:8000/todos/{todo1_id}")
    # This should return 404 because the other user doesn't own this todo
    assert res.status_code == 404
    assert "Todo not found" in res.json()["error"]
    print("✓ Unauthorized access to other user's todo returns 404")
    
    # Create a todo as the other user
    res = session.post("http://localhost:8000/todos", 
                       json={"title": "Other user todo", "description": "Owned by other user"})
    assert res.status_code == 201
    todo_other_id = res.json()["id"]
    print("✓ Created a todo for other user")
    
    # Update the other user's todo
    res = session.put(f"http://localhost:8000/todos/{todo_other_id}", 
                     json={"title": "Updated other's todo", "completed": True})
    assert res.status_code == 200
    assert res.json()["title"] == "Updated other's todo"
    assert res.json()["completed"] == True
    print("✓ Other user can update their own todo")
    
    print("\n=== Testing PASSWORD Change ===")
    # Test change password with old password verification
    res = session.put("http://localhost:8000/password", 
                     json={"old_password": "password123", 
                           "new_password": "newpassword456"})
    assert res.status_code == 200
    print("✓ Password change successful")
    
    # Logout first
    res = session.post("http://localhost:8000/logout")
    assert res.status_code == 200
    
    # Now try logging in with new password
    session.cookies.clear()
    res = session.post("http://localhost:8000/login", 
                      json={"username": "otheruser", "password": "newpassword456"})
    assert res.status_code == 200
    print("✓ Login with new password successful")
    
    # Try changing password with wrong old password
    res = session.put("http://localhost:8000/password", 
                     json={"old_password": "wrongpassword", 
                           "new_password": "anotherpassword"})
    assert res.status_code == 401
    assert "Invalid credentials" in res.json()["error"]
    print("✓ Password change with wrong current password fails")
    
    # Try changing to short password
    res = session.put("http://localhost:8000/password", 
                     json={"old_password": "newpassword456", 
                           "new_password": "short"})
    assert res.status_code == 400
    assert "Password too short" in res.json()["error"]
    print("✓ Password change to short password fails")
    
    print("\n=== Testing DELETE Todo ===")
    # Create a todo to delete
    res = session.post("http://localhost:8000/todos", 
                       json={"title": "ToDelete todo", "description": "Will be deleted"})
    assert res.status_code == 201
    delete_todo_id = res.json()["id"]
    
    # Delete that todo
    res = session.delete(f"http://localhost:8000/todos/{delete_todo_id}")
    assert res.status_code == 204  # 204 means no content per spec
    print("✓ Delete todo successful")
    
    # Confirm deletion
    res = session.get(f"http://localhost:8000/todos/{delete_todo_id}")
    assert res.status_code == 404
    print("✓ Deleted todo confirmed gone")
    
    # Try to delete non-existent todo
    res = session.delete("http://localhost:8000/todos/9999")
    assert res.status_code == 404
    print("✓ Delete non-existent todo returns 404")
    
    print("\n=== Final Validation: All Core Functionality Works ===")
    
    # Shutdown server
    process.terminate()
    try:
        process.wait(timeout=5)
    except subprocess.TimeoutExpired:
        process.kill()
    
    print("\n✓ ALL TESTS PASSED! Server implementation is working correctly.")


if __name__ == "__main__":
    main()